<a href="https://colab.research.google.com/github/mannb986/life_expectancy_nn/blob/main/life_expectancy_nn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Implementing Neural Networks

The World Health Organization (WHO)’s Global Health Observatory (GHO) data repository tracks life expectancy for countries worldwide by following health status and many other related factors.

Although there have been a lot of studies undertaken in the past on factors affecting life expectancy considering demographic variables, income composition, and mortality rates, it was found that the effects of immunization and human development index were not taken into account.

This dataset covers a variety of indicators for all countries from 2000 to 2015 including:

* Immunization factors
* Mortality factors
* Economic factors
* Social factors
* other health-related factors

Ideally, this data will eventually inform countries concerning which factors to change in order to improve the life expectancy of their populations. If we can predict life expectancy well given all the factors, this is a good sign that there are some important patterns in the data. Life expectancy is expressed in years, and hence it is a number. This means that in order to build a predictive model one needs to use regression.

In this project I will design, train and evaluate a neural network model performing the task of regression to predict the life expectancy of countries using this dataset.

## Data Loading & Observing

In [1]:
from google.colab import files


uploaded = files.upload()

Saving life_expectancy.csv to life_expectancy.csv


In [2]:
import pandas as pd

dataset = pd.read_csv('life_expectancy.csv')

In [3]:
dataset.head()

Country  Year  ... Schooling  Life expectancy
0  Afghanistan  2015  ...      10.1             65.0
1  Afghanistan  2014  ...      10.0             59.9
2  Afghanistan  2013  ...       9.9             59.9
3  Afghanistan  2012  ...       9.8             59.5
4  Afghanistan  2011  ...       9.5             59.2

[5 rows x 22 columns]

In [4]:
dataset.describe()

Year  Adult Mortality  ...    Schooling  Life expectancy
count  2938.000000      2938.000000  ...  2938.000000      2938.000000
mean   2007.518720       164.725664  ...    12.009837        69.234717
std       4.613841       124.086215  ...     3.265139         9.509115
min    2000.000000         1.000000  ...     0.000000        36.300000
25%    2004.000000        74.000000  ...    10.300000        63.200000
50%    2008.000000       144.000000  ...    12.300000        72.100000
75%    2012.000000       227.000000  ...    14.100000        75.600000
max    2015.000000       723.000000  ...    20.700000        89.000000

[8 rows x 20 columns]



I will drop the country column, why? To create a predictive model, knowing from which country data comes can be confusing and it is not a column I can generalize over. I want to learn a general pattern for all the countries, and not only those dependent on specific countries.



In [5]:
dataset = dataset.drop(['Country'], axis=1)

In [6]:
labels = dataset.iloc[:,-1]

In [7]:
features = dataset.iloc[:, 0:-1]

In [8]:
features.head()

Year      Status  ...  Income composition of resources  Schooling
0  2015  Developing  ...                            0.479       10.1
1  2014  Developing  ...                            0.476       10.0
2  2013  Developing  ...                            0.470        9.9
3  2012  Developing  ...                            0.463        9.8
4  2011  Developing  ...                            0.454        9.5

[5 rows x 20 columns]

In [9]:
labels.head()

0    65.0
1    59.9
2    59.9
3    59.5
4    59.2
Name: Life expectancy, dtype: float64

## Data Processing

In [10]:
features = pd.get_dummies(features)

In [11]:
features.head()

Year  Adult Mortality  ...  Status_Developed  Status_Developing
0  2015            263.0  ...                 0                  1
1  2014            271.0  ...                 0                  1
2  2013            268.0  ...                 0                  1
3  2012            272.0  ...                 0                  1
4  2011            275.0  ...                 0                  1

[5 rows x 21 columns]

In [21]:
from sklearn.model_selection import train_test_split

features_train, features_test, labels_train, labels_test = train_test_split(features, labels, test_size = 0.33, random_state = 42)

In [22]:
from sklearn.preprocessing import Normalizer
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer 

numerical_features = features.select_dtypes(include=['float64', 'int64'])
numerical_columns = numerical_features.columns

ct = ColumnTransformer([("only numeric", StandardScaler(), numerical_columns)], remainder='passthrough')

In [23]:
features_train_scaled = ct.fit_transform(features_train)

In [24]:
features_test_scaled = ct.transform(features_test)

## Building the Model

In [25]:
from tensorflow.keras.models import Sequential

my_model = Sequential()

In [27]:
from tensorflow.keras.layers import InputLayer

#creating the input layer with the shape corresponding to the number of features
input = InputLayer(input_shape=(features.shape[1],))

In [28]:
#adding the input layer to the model
my_model.add(input)

In [29]:
my_model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
Total params: 0
Trainable params: 0
Non-trainable params: 0
_________________________________________________________________


In [31]:
from tensorflow.keras.layers import Dense

#adding one hidden layer with 64 neurons
my_model.add(Dense(64, activation='relu'))

In [33]:
#adding output layer with a single output
my_model.add(Dense(1))

In [34]:
my_model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 64)                1408      
_________________________________________________________________
dense_1 (Dense)              (None, 64)                4160      
_________________________________________________________________
dense_2 (Dense)              (None, 1)                 65        
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 2         
Total params: 5,635
Trainable params: 5,635
Non-trainable params: 0
_________________________________________________________________


## Initializing the Optimizer & Compiling the Model

In [35]:
from tensorflow.keras.optimizers import Adam

opt = Adam(learning_rate=0.01)

In [37]:
my_model.compile(loss = 'mse', metrics = 'mae', optimizer=opt)

## Fit & Evaluate the Model

In [38]:
my_model.fit(features_train_scaled, labels_train, epochs=40, batch_size=1, verbose=1)

Epoch 1/40
1968/1968 [==============================] - 6s 1ms/step - loss: 433.4377 - mae: 12.9265
Epoch 2/40
1968/1968 [==============================] - 3s 1ms/step - loss: 53.9465 - mae: 5.2627
Epoch 3/40
1968/1968 [==============================] - 3s 1ms/step - loss: 29.8414 - mae: 4.1400
Epoch 4/40
1968/1968 [==============================] - 3s 1ms/step - loss: 34.3734 - mae: 4.1201
Epoch 5/40
1968/1968 [==============================] - 3s 1ms/step - loss: 18.9142 - mae: 3.3285
Epoch 6/40
1968/1968 [==============================] - 3s 1ms/step - loss: 23.3455 - mae: 3.5635
Epoch 7/40
1968/1968 [==============================] - 3s 1ms/step - loss: 19.6253 - mae: 3.2652
Epoch 8/40
1968/1968 [==============================] - 3s 1ms/step - loss: 20.7434 - mae: 3.4015
Epoch 9/40
1968/1968 [==============================] - 3s 1ms/step - loss: 16.3131 - mae: 2.9497
Epoch 10/40
1968/1968 [==============================] - 3s 1ms/step - loss: 14.9690 - mae: 2.9398
Epoch 11/40
1968/

In [39]:
res_mse, rse_mae = my_model.evaluate(features_test_scaled, labels_test, verbose=0)

In [40]:
print("MSE ", res_mse)
print("MAE ", rse_mae)

MSE  6.49599027633667
MAE  1.832485556602478


In [42]:
from sklearn.dummy import DummyRegressor
from sklearn.metrics import mean_absolute_error
dummy_regr = DummyRegressor(strategy="mean")
dummy_regr.fit(features_train_scaled, labels_train)
y_pred = dummy_regr.predict(features_test_scaled)
MAE_baseline = mean_absolute_error(labels_test, y_pred)
print("MAE baseline: ", MAE_baseline)

MAE baseline:  7.833263766658285


## Results

We can see from the output that the model has performed very well with an mean absoulte error of just over 1.8. When compared to a baseline of 7.8, the model has performed very well. 